common/cuda_hip/matrix/csr_kernels.hpp.inc

```diff
@@ -869,37 +813,105 @@ __global__ __launch_bounds__(default_block_size) void add_scaled_identity(
+                                                     thrust::get<1>(b) / bs);
+                        });
+    // build block pattern
+    auto adj_predicate = [bs, in_rows, in_cols, nnz] __device__(size_type i) {
+        const auto a_block_row = i > 0 ? in_rows[i - 1] / bs : -1;
+        const auto a_block_col = i > 0 ? in_cols[i - 1] / bs : -1;
+        const auto b_block_row = in_rows[i] / bs;
+        const auto b_block_col = in_cols[i] / bs;
+        return (a_block_row != b_block_row) || (a_block_col != b_block_col);
+    };
+    auto iota = thrust::make_counting_iterator(size_type{});
+    // count how many blocks we have by counting how often the block changes
+    auto num_blocks = static_cast<size_type>(
+        thrust::count_if(thrust::device, iota, iota + nnz, adj_predicate));
+    // allocate storage
+    Array<IndexType> block_row_idx_array{exec, num_blocks};
+    Array<size_type> block_ptr_array{exec, num_blocks};
+    block_col_idx_array.resize_and_reset(num_blocks);
+    block_value_array.resize_and_reset(num_blocks * bs * bs);
+    auto row_idxs = block_row_idx_array.get_data();
+    auto col_idxs = block_col_idx_array.get_data();
+    auto values = as_device_type(block_value_array.get_data());
+    auto block_ptrs = block_ptr_array.get_data();
+    auto block_ptr_it = thrust::device_pointer_cast(block_ptrs);
+    // write (block_row, block_col, block_start_idx) tuples for each block
+    thrust::copy_if(thrust::device, iota, iota + nnz, block_ptr_it,
+                    adj_predicate);
+    auto block_output_it = thrust::make_zip_iterator(
+        thrust::make_tuple(thrust::device_pointer_cast(row_idxs),
+                           thrust::device_pointer_cast(col_idxs)));
+    thrust::transform(
+        thrust::device, block_ptr_it, block_ptr_it + num_blocks,
+        block_output_it, [bs, in_rows, in_cols] __device__(size_type i) {
+            return thrust::make_tuple(in_rows[i] / bs, in_cols[i] / bs);
+        });
+    // build row pointers from row indices
+    components::convert_idxs_to_ptrs(exec, block_row_idx_array.get_const_data(),
+                                     block_row_idx_array.get_num_elems(),
+                                     num_block_rows, block_row_ptrs);
+    // fill in values
+    components::fill_array(exec, block_value_array.get_data(),
+                           num_blocks * bs * bs, zero<ValueType>());
+    thrust::for_each_n(
+        thrust::device, iota, num_blocks,
+        [block_ptrs, nnz, num_blocks, bs, in_rows, in_cols, in_vals,
+         values] __device__(size_type i) {
+            const auto block_begin = block_ptrs[i];
+            const auto block_end = i < num_blocks - 1 ? block_ptrs[i + 1] : nnz;
+            for (auto nz = block_begin; nz < block_end; nz++) {
+                values[i * bs * bs + (in_cols[nz] % bs) * bs +
+                       (in_rows[nz] % bs)] = fake_complex_unpack(in_vals[nz]);
+            }
+        });
```

Comment:
this would require a lookup nonzero -> block index, which we don't have at that point. We can replace this by our own warp-parallel kernel in the future, where we can take care of load-balancing between blocks, and doing (almost) coalescing reads.